### PR TITLE
[improvement] Helpful error output is provided when formatting fails

### PR DESCRIPTION
**Pull request by carterkozak**
## Before this PR
```
java.lang.RuntimeException: com.google.googlejavaformat.java.FormatterException: 5:15: error: ';' expected

	at com.palantir.conjure.java.util.Goethe.formatAndEmit(Goethe.java:55)
	at com.palantir.conjure.java.util.GoetheTest.testFormatAndEmit(GoetheTest.java:38)
Caused by: com.google.googlejavaformat.java.FormatterException: 5:15: error: ';' expected
	at com.google.googlejavaformat.java.FormatterException.fromJavacDiagnostics(FormatterException.java:50)
	at com.google.googlejavaformat.java.Formatter.format(Formatter.java:151)
	at com.google.googlejavaformat.java.Formatter.getFormatReplacements(Formatter.java:258)
	at com.google.googlejavaformat.java.Formatter.formatSource(Formatter.java:234)
	at com.google.googlejavaformat.java.Formatter.formatSource(Formatter.java:202)
	at com.google.googlejavaformat.java.Formatter.formatSource(Formatter.java:189)
	at com.palantir.conjure.java.util.Goethe.formatAndEmit(Goethe.java:52)
	... 25 more
```

## After this PR
```
java.lang.RuntimeException: Failed to format 'com.palantir.foo.Foo'
';' expected
    type oops name = bar;
            ^

	at com.palantir.conjure.java.util.Goethe.formatAndEmit(Goethe.java:55)
	at com.palantir.conjure.java.util.GoetheTest.testFormatAndEmit(GoetheTest.java:38)
Caused by: com.google.googlejavaformat.java.FormatterException: 5:15: error: ';' expected
	at com.google.googlejavaformat.java.FormatterException.fromJavacDiagnostics(FormatterException.java:50)
	at com.google.googlejavaformat.java.Formatter.format(Formatter.java:151)
	at com.google.googlejavaformat.java.Formatter.getFormatReplacements(Formatter.java:258)
	at com.google.googlejavaformat.java.Formatter.formatSource(Formatter.java:234)
	at com.google.googlejavaformat.java.Formatter.formatSource(Formatter.java:202)
	at com.google.googlejavaformat.java.Formatter.formatSource(Formatter.java:189)
	at com.palantir.conjure.java.util.Goethe.formatAndEmit(Goethe.java:52)
	... 25 more
```

==COMMIT_MSG==
Helpful error output is provided when formatting fails
==COMMIT_MSG=

### CODE REVIEW

File: conjure-java-core/src/test/java/com/palantir/conjure/java/util/GoetheTest.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    @Test
+    public void testFormatAndEmit() {
+        JavaFile javaFile = JavaFile.builder("com.palantir.foo", TypeSpec.classBuilder("Foo")
+                .addStaticBlock(CodeBlock.builder().addStatement("type oops name = bar").build())
+                .build()).build();
+        assertThatThrownBy(() -> Goethe.formatAndEmit(javaFile, folder.getRoot().toPath()))
+                .hasMessageContaining("Failed to format 'com.palantir.foo.Foo'")
+                .hasMessageContaining("';' expected")
+                .hasMessageContaining("" // newline to align the output
+                        + "    type oops name = bar;\n"
```
Comment (rubenfiszel) [inline conjure-java-core/src/test/java/com/palantir/conjure/java/util/GoetheTest.java:42]: nit: is there any way to remove the whitespace at the beginning of this line?

Comment (carterkozak) [reply conjure-java-core/src/test/java/com/palantir/conjure/java/util/GoetheTest.java:42]: We would have to count it, and subtract off of the column counter as well, seemed more complicated than it was worth :-)

**Review by rubenfiszel — APPROVED** [review]

File: conjure-java-core/src/main/java/com/palantir/conjure/java/util/Goethe.java [hunk 2]
```diff
@@ -40,19 +44,44 @@ private Goethe() {}
...
+            StringBuilder failureText = new StringBuilder();
+            failureText.append("Failed to format '")
+                    .append(file.packageName).append('.').append(file.typeSpec.name).append("'\n");
+            for (FormatterDiagnostic formatterDiagnostic : formatterDiagnostics) {
+                failureText.append(formatterDiagnostic.message()).append("\n")
+                        // Diagnostic values are one-indexed, while our list is zero-indexed.
+                        .append(lines.get(formatterDiagnostic.line() - 1))
+                        .append('\n')
+                        // Offset by two to convert from one-indexed to zero indexed values, and account for the
+                        // carrot.
```
Comment (dansanduleac) [inline conjure-java-core/src/main/java/com/palantir/conjure/java/util/Goethe.java]: caret? 🥕

Comment (carterkozak) [reply conjure-java-core/src/main/java/com/palantir/conjure/java/util/Goethe.java]: 🤦‍♂

File: conjure-java-core/src/main/java/com/palantir/conjure/java/util/Goethe.java [hunk 3]
```diff
@@ -40,19 +44,44 @@ private Goethe() {}
...
+                        .append(Strings.repeat(" ", Math.max(0, formatterDiagnostic.column() - 2)))
+                        .append("^\n\n");
```
Comment (dansanduleac) [inline conjure-java-core/src/main/java/com/palantir/conjure/java/util/Goethe.java:78]: this seems fine but i'm surprised they don't provide this kind of diagnostic printer

Comment (carterkozak) [reply conjure-java-core/src/main/java/com/palantir/conjure/java/util/Goethe.java:78]: Me too, there are no code references to the line/column values, only printing diagnostic error message strings.

File: conjure-java-core/src/main/java/com/palantir/conjure/java/util/Goethe.java [hunk 4]
```diff
@@ -40,19 +44,43 @@ private Goethe() {}
...
             Path outputFile = getFilePath(file, baseDir);
-            StringBuilder code = new StringBuilder();
             file.writeTo(code);
 
             CharSink sink = com.google.common.io.Files.asCharSink(outputFile.toFile(), StandardCharsets.UTF_8);
             JAVA_FORMATTER.formatSource(CharSource.wrap(code), sink);
             return outputFile;
-        } catch (IOException | FormatterException e) {
+        } catch (FormatterException e) {
+            throw new RuntimeException(generateMessage(file, code.toString(), e.diagnostics()), e);
```
Comment (dansanduleac) [inline conjure-java-core/src/main/java/com/palantir/conjure/java/util/Goethe.java]: can we split this catch block to only apply to the JAVA_FORMATTER bit?
seems weird that this catch block relies on getting past `file.writeTo(code)` inside the same try block.

Comment (carterkozak) [reply conjure-java-core/src/main/java/com/palantir/conjure/java/util/Goethe.java]: ack, updated.

**Review by dansanduleac — APPROVED** [review]

**Pull request merged**